cartographer_ros/cartographer_ros/map_builder_bridge.cc

```diff
@@ -242,6 +244,113 @@ visualization_msgs::MarkerArray MapBuilderBridge::GetTrajectoryNodesList() {
+  constraint_intra_marker.scale.x = kConstraintLineListMarkerScale;
+  constraint_intra_marker.pose.orientation.w = 1.0;
+
+  visualization_msgs::Marker residual_intra_marker = constraint_intra_marker;
+  residual_intra_marker.id = marker_id++;
+  residual_intra_marker.ns = "Intra residuals";
+  residual_intra_marker.pose.position.z = 0.1;
+
+  visualization_msgs::Marker constraint_inter_marker = constraint_intra_marker;
+  constraint_inter_marker.id = marker_id++;
+  constraint_inter_marker.ns = "Inter constraints";
+  constraint_inter_marker.pose.position.z = 0.1;
+
+  visualization_msgs::Marker residual_inter_marker = constraint_intra_marker;
+  residual_inter_marker.id = marker_id++;
+  residual_inter_marker.ns = "Inter residuals";
+  residual_inter_marker.pose.position.z = 0.1;
+
+  const auto all_trajectory_nodes =
+      map_builder_.sparse_pose_graph()->GetTrajectoryNodes();
+  const auto all_submap_data =
+      map_builder_.sparse_pose_graph()->GetAllSubmapData();
+  const auto constraints = map_builder_.sparse_pose_graph()->constraints();
+
+  // Color mapping for submaps of various trajectories - add trajectory id
+  // to ensure different starting colors.
+  const auto submap_color_id_map =
+      [](const cartographer::mapping::SubmapId submap_id) -> int {
+    return submap_id.submap_index + submap_id.trajectory_id + 20;
+  };
+
+  for (const auto& constraint : constraints) {
+    visualization_msgs::Marker *constraint_marker, *residual_marker;
+    std_msgs::ColorRGBA color_constraint, color_residual;
+    if (constraint.tag ==
+        cartographer::mapping::SparsePoseGraph::Constraint::INTRA_SUBMAP) {
+      constraint_marker = &constraint_intra_marker;
+      residual_marker = &residual_intra_marker;
+      color_constraint = GetColor(submap_color_id_map(constraint.submap_id));
+      color_residual.a = 1.0;
+      color_residual.r = 1.0;
+    } else {
+      constraint_marker = &constraint_inter_marker;
+      residual_marker = &residual_inter_marker;
+      // yellow
+      color_constraint.a = 1.0;
+      color_constraint.r = color_constraint.g = 1.0;
+      // cyan
+      color_residual.a = 1.0;
+      color_residual.b = color_residual.g = 1.0;
+    }
+
+    constraint_marker->colors.push_back(color_constraint);
+    constraint_marker->colors.push_back(color_constraint);
+    residual_marker->colors.push_back(color_residual);
+    residual_marker->colors.push_back(color_residual);
+
+    const auto& submap_data =
+        all_submap_data[constraint.submap_id.trajectory_id]
+                       [constraint.submap_id.submap_index];
+    const auto& submap_origin = submap_data.pose;
+    const auto& trajectory_node_pose =
+        all_trajectory_nodes[constraint.node_id.trajectory_id]
+                            [constraint.node_id.node_index]
+                                .pose;
+    const auto constraint_pose = submap_origin * constraint.pose.zbar_ij;
```

Comment:
Let's not use `auto` in these cases. It works now, but I feel it is to easy to miss cases using Eigen types.